custom tip format

mbostock opened the pull request:
This implements a new **format** option for the tip mark that specifies how to format values. The option is specified as an object whose property names are channel names and whose values are formats; each format may either be a shorthand string for a number or time format, or a function that takes a channel value and zero-based index and returns the corresponding formatted string.

For example, to customize the number format of the *x* channel:

```js
format: {x: ".2f"}
```

Equivalently:

```js
format: {x: (d) => d.toFixed(2)}
```

To suppress a channel (similar to #1822), set the corresponding format to null:

```js
format: {x: null}
```

As before, if you want the tip to display a paragraph of text, you can use the **title** channel; if you want to add additional name-value pairs to the tip, you can declare extra channels using the **channels** option; if you want to change the label, you can either change the corresponding channel or scale label. This PR improves the **channels** option slightly, allowing the **label** option to be declared alongside a channel #1734, and defaulting the channel label to its name when expressed as a shorthand string (e.g., `channels: {Sport: "sport"}` now uses the label “Sport” instead of “sport”.)

Lastly, the **tip** mark option now supports passing arbitrary options through to the derived tip mark, including the **format** option. The **tip** mark option also now takes a **pointer** option to control which pointer mode is used (_x_, _y_, or _xy_, for pointerX, pointerY, or pointer respectively).

Fixes #1612.
Fixes #1733.
Fixes #1734.

Fil commented:
Love the API design.

color, opacity… can we have symbol too, maybe? :-)

mkfreeman commented:
Very exciting! Would it be possible to pass up the `format` option from other marks in the `inferTip` [function](https://github.com/observablehq/plot/blob/main/src/plot.js#L523)? The first thing I tried to do was use a tip formatter with another mark where `tip: true`. 

I think it would be a big win for new-ish users to be able to format their tips without ejecting from `tip: true` to adding another tip mark (might also make it easier to add the UI in the chart cell?).

mbostock commented:
Yes @mkfreeman; see the top description.

mkfreeman commented:
Does it make sense to put the swatch at the beginning of each line (to match legends)? Just wanted to compare:

| Start  | Middle | end |
| ------------- | ------------- | ------------- |
| <img width="144" alt="beginning" src="https://github.com/observablehq/plot/assets/1373882/a1b1dc09-fd06-4abd-9853-e48ba57ecf1b"> |<img width="146" alt="middle" src="https://github.com/observablehq/plot/assets/1373882/fc2178c7-24ee-454e-bb67-54deb5dbe9e2"> |  <img width="147" alt="end" src="https://github.com/observablehq/plot/assets/1373882/b56ae3cb-0010-4f6e-8934-e624a146db95">|

mbostock commented:
@mkfreeman I’m not changing the display functionality of tips in this PR; I’m just giving more control how the contents of the tip are specified. See #1767 for a more general issue, but feel free to open another issue if you want to suggest changes to the default design of tips.

mkfreeman commented:
@mbostock thanks, will take a look at that issue!

mbostock commented:
I’m wondering if I over-engineered this by adding the ability to derive anonymous channels as part of the format. I could do something that is more focused on just how the channels are formatted. For example, instead of this:

```js
format: [{channel: "x", format: ".2f"}]
```

maybe you’d say:

```js
format: {x: ".2f"}
```

and it wouldn’t change which channels were displayed, but it would change how the _x_ channel is formatted (or perhaps how any channel bound to the _x_ scale is formatted? not sure). And maybe you could suppress a channel by saying:

```js
format: {x: ".2f", y: null}
```

I think that might be enough. 🤔

mbostock commented:
After much exploration, I’ve decided to pare this PR down to its essentials, focusing solely on formatting channel values and suppressing channels. The other functionality (declaring additional channels, and overriding labels) is already covered by the existing API, so I didn’t want to introduce a competing mechanism.

mbostock commented on src/marks/tip.js:
```diff
@@ -332,18 +334,73 @@ function getSources({channels}) {
+      };
+    } else {
+      const format = this.format?.[key];
+      if (format === null) continue;
```
Possible future improvement: we could prune the result of getSources based on this.format; then we wouldn’t need to skip here. Also, we could materialize the formatDefault as needed.

Fil commented on src/plot.js:
```diff
@@ -239,6 +239,7 @@ export function plot(options = {}) {
   // Compute value objects, applying scales and projection as needed.
   for (const [mark, state] of stateByMark) {
     state.values = mark.scale(state.channels, scales, context);
+    state.values.data = state.data; // expose transformed data for advanced usage
```
It will be immensely useful in the render API to always have access to mark’s (arrayified, possibly transformed) data. I just wonder why it appears here, as it seems unrelated to this PR. Needs to be added to the documentation #1811.

mbostock replied:
This was needed in an earlier variation of this PR where you could specify the format as a function of data. I don’t think it’s needed anymore, so I can remove it and we can add it as a separate PR when we are ready.

Fil commented:
Great!! Besides the missing documentation, two minor (non-blocking) remarks:

1. I was tempted to say `format: {x: true} `to just use the standard format (for instance, when I wanted to revert `{x: null}`).


2. Looks like this enables to set the order of the fields by setting the order in the channels options, and possibly suppressing the corresponding standard channels. For example:

```
Plot.dot(penguins, {
  x: "body_mass_g",
  y: "culmen_length_mm",
  fill: "species",
  channels: {
    culmen_length_mm: "culmen_length_mm",
    species: { value: "species", label: "species", scale: "color" },
    body_mass_g: "body_mass_g"
  },
  tip: { format: { fill: null, culmen_length_mm: ".1f", x: null } }
}).plot()
```

results in the order (culmen_length_mm, species, body_mass_g)

<img width="242" alt="Capture d’écran 2023-08-25 à 11 07 14" src="https://github.com/observablehq/plot/assets/7001/1cb30f3f-886d-44c4-a2fb-1a852e93a4a8">

I feel that this is a bit arbitrary and hacky (and also bad performance-wise). Since it's something that I think users will want to do anyways, it would be nicer to design this such that it followed the order given in the _format_ instead. This would be an additional reason to support `{x: true}`, btw.

Fil commented:
To expand my remark about symbols: it would be great to have a symbol (where there is a swatch) for channels that use the symbol scale. And the combo color+symbol should apply too. Again not a blocker, I can try to do this separately.

mbostock commented:
Let’s do the symbol swatch separately. 

The order of channels is hard and not addressed in this PR. The formats are sparse and I think using “true” as a format is weird — that doesn’t work for a tickFormat, for example. I think channel order might need to be a separate PR or option too, but I’ll think about it today. (I’ve already spent a lot of time on this and want to move forward to other things for 1.0.)

mbostock commented:
I could change it to do whatever is listed in the **format** option first, followed by any remaining channels (in arbitrary order). That seems like a reasonable improvement although it might complicate the code some. We will see! I can probably address the improvement I mentioned above at the same time. But then I want to move on!

mbostock commented:
The **format** option now determines the order in which channels are shown, and you can set the format to true (or equivalently undefined) if you just want to use it to specify the order. I also added the optimization I described above to materialize the defaults, fixed a bug with facet settings (**facet** and **facetAnchor**) for the **tip** mark option, and removed the _data_ pseudochannel.

Still need documentation and tests, I guess. Maybe that could be a followup PR? 😬

mbostock commented on src/marks/tip.js:
```diff
@@ -322,28 +327,86 @@ function getPath(anchor, m, r, width, height) {
+    if (key === "x2" && "x1" in channels) {
+      yield {
+        label: formatPairLabel(scales, channels, "x"),
+        value: formatPair(this.format.x, channels.x1, channel, i)
```
I think this might approach not work now, at least if you use a shorthand string for the _x_ or _y_ format, because the name of the format (*x*) does not correspond to the name of the channel (*x1* and *x2*). I could switch to using the *x2* here, but I think supporting *x* is more natural since the same format is used for both channels. 🤔

Fil replied:
Ah yes this breaks the tipAreaBand, tipAreaStack tests (and maybe others), because format.y is undefined. (Too bad we can't catch those automagically.)

Fil replied:
I need to run, but here's a fix:

```patch
diff --git a/src/marks/tip.js b/src/marks/tip.js
index 0a271e1d..97fa52bb 100644
--- a/src/marks/tip.js
+++ b/src/marks/tip.js
@@ -372,12 +372,12 @@ function* formatChannels(i, index, channels, scales, values) {
     if (key === "x2" && "x1" in channels) {
       yield {
         label: formatPairLabel(scales, channels, "x"),
-        value: formatPair(this.format.x, channels.x1, channel, i)
+        value: formatPair(this.format.x1 ?? this.format.x, this.format.x2 ?? this.format.x, channels.x1, channel, i)
       };
     } else if (key === "y2" && "y1" in channels) {
       yield {
         label: formatPairLabel(scales, channels, "y"),
-        value: formatPair(this.format.y, channels.y1, channel, i)
+        value: formatPair(this.format.y1 ?? this.format.y, this.format.y2 ?? this.format.y, channels.y1, channel, i)
       };
     } else {
       const value = channel.value[i];
@@ -393,10 +393,10 @@ function* formatChannels(i, index, channels, scales, values) {
   }
 }

-function formatPair(formatValue, c1, c2, i) {
+function formatPair(formatValue1, formatValue2, c1, c2, i) {
   return c2.hint?.length // e.g., stackY’s y1 and y2
-    ? `${formatValue(c2.value[i] - c1.value[i], i)}`
-    : `${formatValue(c1.value[i], i)}–${formatValue(c2.value[i], i)}`;
+    ? `${formatValue1(c2.value[i] - c1.value[i], i)}`
+    : `${formatValue1(c1.value[i], i)}–${formatValue2(c2.value[i], i)}`;
 }

 function formatPairLabel(scales, channels, key) {

```

It seems to work on tipAreaStack and tipBin (I can format separately x1 and x2). I haven't investigated what it gives on tips that show a length/difference instead of an extent (?).

mbostock replied:
I pushed what I think is a more general fix, allowing you to specify a format for the _x_ channel as shorthand for both _x1_ and _x2_ (and same for _y_). This means you can control the order in which paired channels appear, and even suppress one of the two paired channels to just show the other. For example, `format: {y1: false}` will just show the high:

<img width="625" alt="Screenshot 2023-08-25 at 12 52 59 PM" src="https://github.com/observablehq/plot/assets/230541/ffcb7428-8a1c-427e-bd14-21028d8378a4">

And `format: {y: true}` will give _y1_ and _y2_ higher priority:

<img width="640" alt="Screenshot 2023-08-25 at 12 53 37 PM" src="https://github.com/observablehq/plot/assets/230541/c8573f7c-968d-4b11-8c27-f370e73d0b3a">

mbostock commented on src/marks/tip.js:
```diff
@@ -322,28 +307,125 @@ function getPath(anchor, m, r, width, height) {
+      if (key === "x" || key === "y") {
+        const key1 = `${key}1`;
+        const key2 = `${key}2`;
+        if (!(key1 in format) && (source = getSource(channels, key1))) (format[key1] = value), (sources[key1] = source);
+        if (!(key2 in format) && (source = getSource(channels, key2))) (format[key2] = value), (sources[key2] = source);
```
Ugh, this only works the first time the mark is rendered. Because we mutate `format` here, thus adding the paired channels to the end, we’re changing the order the second time around. 🤦 I need to do this in a way that puts the paired channels (_x1_ and _x2_) at the same priority as the shorthand (_x_).

dperetti commented:
Is it supposed to work for default channels? I'm not able to suppress / reorder default those, only works for extra channels.

mbostock commented:
@dperetti Yes, it does. Please open a discussion here or in the Observable Forum if you’d like help, and please include a live code example to demonstrate. Thank you.

dperetti commented:
Thanks, finally got it! Turns out the channels were `x` and `y`, not the keys in the data object. This was too much convolution for my aging brain.
